### ISSUE TITLE
Is there a way to perform some action right before completer (or) on KeyPress?

### ISSUE THREAD

**Issue by Shriram-Balaji** [kkawakam/rustyline#441]
Hey,

I'd like to know if there's a way to catch a `TAB` key press, and then populate the `ReadlineHelper` struct with possible completions. 

In particular, I'm trying to implement REPL tab completion for JSTime [here](https://github.com/Shriram-Balaji/jstime/blob/repl-tab-completion/cli/main.rs).

For example, after [setting the helper](https://github.com/Shriram-Balaji/jstime/blob/c21d87d107a3a2e2f50a8a2b61605da1a3b869f1/cli/main.rs#L182) and right before the completer starts to the `complete` method, I'd like to basically fill in the [`completions` in the struct](https://github.com/Shriram-Balaji/jstime/blob/c21d87d107a3a2e2f50a8a2b61605da1a3b869f1/cli/main.rs#L178).

I'm only starting to get familiar with Rust, so I might have missed something here as well :)

**Comment by bjorn3** [comment]
I think you are supposed to calculate the completions directly in the `complete` method of the helper instead of precalculating them and storing them in a field of the helper.

**Comment by Shriram-Balaji** [comment]
@bjorn3 So I was intending to do that, but there some problems I faced with that. For eg. the signature for the `complete` method takes in `&self`, whereas the function that generates these completions requires a `&mut self`.

I could not find a direct way around this, because the trait signatures would be different if I modified the complete method.

**Comment by bjorn3** [comment]
You can use `RefCell` to wrap whatevee you need to mutate.

**Comment by gwenn** [comment]
I agree that the perfect [signature](https://docs.rs/rustyline/6.3.0/rustyline/completion/trait.Completer.html#method.complete) should have been:
```rust
fn complete(
    &mut self, -- mutable
...
```
It appears that all similar crates have the same issue:
* https://docs.rs/linefeed/0.6.0/linefeed/complete/trait.Completer.html#tymethod.complete
* https://docs.rs/liner/0.4.4/liner/trait.Completer.html
* https://docs.rs/termwiz/0.9.0/termwiz/lineedit/trait.LineEditorHost.html#method.complete

For the moment, do as suggested by @bjorn3
